KT-48232: Implement kotlin.math.cbrt() (cubic roots)

romainguy opened the pull request:
`cbrt()` is the standard cubic root function that provides several advantages over `pow(x, 1.0/3.0)`:
- Better precision
- Faster
- Behaves better with negative values

The latter is important in some cases, for instance when implementing algorithms such as the [Oklab (https://bottosson.github.io/posts/oklab/) color conversion functions.

The change provides implementations and tests for JVM, JS, WASM, Native.

Fixes [KT-48232](https://youtrack.jetbrains.com/issue/KT-48232).

romainguy commented on libraries/kotlin.test/js/it/build.gradle:
```diff
@@ -52,7 +52,7 @@ tasks.register("populateNodeModules", Copy) {
 }
 
 node {
-    version = '12.18.0'
```
Uploaded by mistake, will fix.

ilya-g reviewed (commented):
Thank you, this PR is impressive in its thoroughness.

We can take these functions in the coming Kotlin 1.7.20 as experimental and then stabilize later in 1.8. For that they have to be annotated appropriately: please add `@SinceKotlin("1.7")` and `@ExperimentalStdlibApi` annotations to them.

romainguy commented on libraries/stdlib/test/numbers/MathTest.kt:
```diff
@@ -121,6 +121,41 @@ class DoubleMathTest {
     }
 
+    @Test fun cubeRoots() {
+        assertEquals(3.0, cbrt(27.0))
```
👍 

Re: more numbers. Would adding a few numbers in upper ranges and very close to 0 be what you're looking for?

romainguy commented on kotlin-native/runtime/src/main/kotlin/kotlin/math/math.kt:
```diff
@@ -384,6 +384,21 @@ public actual fun max(a: Double, b: Double): Double = when {
+ *   - If the argument is zero, then the result is a zero with the same sign as the argument.
+ */
+@GCUnsafeCall("Kotlin_math_cbrt")
+external public actual fun cbrt(x: Double): Double
```
Will do.

romainguy commented on libraries/stdlib/common/src/kotlin/MathH.kt:
```diff
@@ -354,6 +354,20 @@ public expect fun min(a: Double, b: Double): Double
 
+
+/**
+ * Returns the cube root of [x]. For positive finite `x`, `cbrt(-x) == -cbrt(x)`;
```
It won't for negative values: `cbrt(-x)` where `x == -27` will be `cbrt(27)` which is not equal to `-cbrt(27)`.

ilya-g replied:
It will be, however, equal to `-cbrt(-27)`, i.e. `-cbrt(x)` where `x = -27`

romainguy replied:
Ah right :) Yes, it will hold for any x, except `NaN`

romainguy commented:
Thanks @ilya-g! I've addressed all your comments I believe.

ilya-g commented on libraries/stdlib/test/numbers/MathTest.kt:
```diff
@@ -121,6 +121,30 @@ class DoubleMathTest {
+        )
+
+        for ((x, result) in testingPairs) {
+            assertEquals(result, cbrt(x), 1e-118)
```
I don't think that the absolute tolerance is suitable here. For example, it fails in test for `kotlin-stdlib-js` with "Expected <0.9654893846056297> with absolute tolerance <1e-118>, actual <0.9654893846056298>".

romainguy replied:
I was afraid of that once the tests would run on a different machine. I'll change to a safer tolerance.

romainguy replied:
I like that idea a lot, I'll give it a try. Thanks for the suggestion!

romainguy commented:
@ilya-g I added your suggestion but Kotlin JS doesn't have a `Float.ulp`. To work around this I added a custom implementation of `Float.ulp` in the common test but I'm not sure if that's what you'd like to see. The best solution I can think of would be to also implement `Float.ulp` in JS as part of this change. What do you think?

ilya-g reviewed (commented):
> Kotlin JS doesn't have a Float.ulp. To work around this I added a custom implementation of Float.ulp in the common test

Ah, yes, that's because we're not happy with the current representation of Float in JS (which is basically the same as Double), so we don't want to provide some bit representation-related functions for Float now, such as `nextUp/Down`, `ulp`, because they either would produce results different from other platforms or will experience a breaking change later, when Float range will be enforced ([KT-24975](https://youtrack.jetbrains.com/issue/KT-24975)).

I think having such `ulpCommon` for Float would be fine for the purposes of this test.

romainguy commented on libraries/stdlib/test/numbers/MathTest.kt:
```diff
@@ -140,7 +154,7 @@ class DoubleMathTest {
 
         for ((x, result) in testingPairs) {
-            assertEquals(result, cbrt(x), 1e-15)
+            assertEquals(result, cbrt(x), if (x.isFinite()) 2.0 * x.ulp else 0.0)
```
Good catch. I really need to brush up on my IEE754 skills!
